openedx/core/djangoapps/safe_sessions/middleware.py

```diff
@@ -469,45 +473,70 @@ def _verify_user(request, response, userid_in_session):
-                                .format('\n'.join(request.debug_user_changes)))
+                extra_logs = []
+
+                # One of the possibilities for why we get an occasional
+                # safe-session user mismatch is that the wrong session is
+                # retrieved from cache. This additional logging should reveal
+                # any such mismatch (without revealing the actual session ID
+                # in logs).
+                session_hashes = [
+                    ('cookie_field', obscure_token(request.session_cookie_string[2:34])),
+                    ('at_request', obscure_token(request.safe_cookie_verified_session_id)),
+                    ('at_response', obscure_token(request.session.session_key)),
+                ]
 
                 session_id_changed = hasattr(request.session, 'session_key') and\
                     request.safe_cookie_verified_session_id != request.session.session_key
+
                 # delete old session id for security
                 del request.safe_cookie_verified_session_id
+                del request.session_cookie_string
+
+                extra_logs.append('Session changed.' if session_id_changed else 'Session did not change.')
+
+                # Allow comparing session IDs in both logs and metrics
+                extra_logs.append(
+                    "Hash of session ID from various sources: " +
+                    '; '.join(f'{k}={v}' for (k, v) in session_hashes)
+                )
+                for source_name, id_hash in session_hashes:
+                    set_custom_attribute(f'safe_sessions.session_id_hash.{source_name}', id_hash)
 
-                log_suffix = 'Session changed.' if session_id_changed else 'Session did not change.'
+                if hasattr(request, 'debug_user_changes'):
+                    extra_logs.append(
+                        'An unsafe user transition was found. It either needs to be fixed or exempted.\n' +
+                        '\n'.join(request.debug_user_changes)
+                    )
 
                 if request_user_object_mismatch and not session_user_mismatch:
                     log.warning(
                         (
                             "SafeCookieData user at initial request '{}' does not match user at response time: '{}' "
-                            "for request path '{}'. {}"
+                            "for request path '{}'.\n{}"
                         ).format(  # pylint: disable=logging-format-interpolation
-                            request.safe_cookie_verified_user_id, request.user.id, request.path, log_suffix
+                            request.safe_cookie_verified_user_id, request.user.id, request.path, '\n'.join(extra_logs)
                         ),
                     )
                     set_custom_attribute("safe_sessions.user_mismatch", "request-response-mismatch")
```

Comment:
@timmc-edx: Maybe for a follow-up ticket/PR, but these message and custom attribute code for user mismatches could possibly also use similar "source" based metrics and comparison. For example:
```
set_custom_attribute(f'safe_sessions.user_id.{source_name}', user_id)
```
In both cases, we might want a separate call-out for if there is a mismatch, which could be True/False. For example:
```
set_custom_attribute("safe_sessions.user_mismatch", user_id_changed)
```

Reply:
Yeah, might be good for a followup. Honestly there's a lot of cleanup/reorganizing I'd like to do overall in this code.